src/XrdVoms/XrdVomsMapfile.cc

```diff
@@ -0,0 +1,423 @@
+                else if (!strcmp(val, "warning")) {erp->setMsgMask(erp->getMsgMask() | LogMask::Warning);}
+                else if (!strcmp(val, "info")) {erp->setMsgMask(erp->getMsgMask() | LogMask::Info);}
+                else if (!strcmp(val, "debug")) {erp->setMsgMask(erp->getMsgMask() | LogMask::Debug);}
+                else if (!strcmp(val, "none")) {erp->setMsgMask(0);}
+                else {erp->Emsg("Config", "voms.trace encountered an unknown directive:", val);}
+                val = stream.GetWord();
+            } while (val);
+        }
+    }
+
+    if (!map_filename.empty()) {
+        if (erp) erp->Emsg("Config", "Will initialize VOMS mapfile", map_filename.c_str());
+        mapper.reset(new XrdVomsMapfile(erp, map_filename));
+        if (!mapper->IsValid()) {
+            mapper.reset(nullptr);
+            return VOMS_MAP_FAILED;
+        }
+    }
+
+    return mapper.get();
+}
+
+
+void *
+XrdVomsMapfile::MaintenanceThread(void *myself_raw)
+{
+    auto myself = static_cast<XrdVomsMapfile*>(myself_raw);
+
+   auto now = monotonic_time_s();
+   auto next_update = now + m_update_interval;
+   while (true) {
+       now = monotonic_time_s();
+       auto remaining = next_update - now;
+       auto rval = sleep(remaining);
+       if (rval > 0) {
+           // Woke up early due to a signal; re-run prior logic.
+           continue;
+       }
+       next_update = monotonic_time_s() + m_update_interval;
+       struct stat statbuf;
+       if (-1 == stat(myself->m_mapfile.c_str(), &statbuf)) {
+           myself->m_edest->Emsg("XrdVomsMapfile", errno, "Error checking the mapfile",
+               myself->m_mapfile.c_str());
+           myself->m_mapfile_ctime.tv_sec = 0;
+           myself->m_mapfile_ctime.tv_nsec = 0;
+           myself->m_is_valid = false;
+           continue;
+       }
+       // Use ctime here as it is solely controlled by the OS (unlike mtime,
+       // which can be manipulated by userspace and potentially not change
+       // when updated - rsync, tar, and rpm, for example, all preserve mtime).
+       // ctime also will also be updated appropriately for overwrites/renames,
+       // allowing us to detect those changes as well.
+       //
+       if ((myself->m_mapfile_ctime.tv_sec == statbuf.st_ctim.tv_sec) &&
+           (myself->m_mapfile_ctime.tv_nsec == statbuf.st_ctim.tv_nsec))
+       {
+           myself->m_edest->Log(LogMask::Debug, "Maintenance", "Not reloading VOMS mapfile; "
+               "no changes detected.");
+           continue;
+       }
+       memcpy(&myself->m_mapfile_ctime, &statbuf.st_ctim, sizeof(decltype(statbuf.st_ctim)));
+
+       myself->m_edest->Log(LogMask::Debug, "Maintenance", "Reloading VOMS mapfile now");
+       if ( !(myself->m_is_valid = myself->ParseMapfile(myself->m_mapfile)) ) {
+           myself->m_edest->Log(LogMask::Error, "Maintenance", "Failed to reload VOMS mapfile");
+       }
+   }
```

Comment:
Apparently we neeed a return statement here. Suggestion is

return nullptr;

See comment from Giuseppe.

Reply:
Addressed in latest commits.